src/lib/expression/abstract_expression.hpp

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "all_type_variant.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+enum class ExpressionType {
+  Aggregate,
+  Arithmetic,
+  Case,
+  Column,
+  Exists,
+  Extract,
+  Function,
+  Negate,
+  List,
+  Logical,
+  Parameter,
+  Predicate,
+  Select,
+  Value
+};
+
+/**
+ * AbstractExpression is the self-contained data structure describing Expressions in Hyrise.
+ *
+ * Expressions in Hyrise are everything down from Literals and Columns, over Arithmetics (a + b, ...),
+ * Logicals (a AND b), up to Lists (`('a', 'b')`) and Subselects. Check out the classes derived from AbstractExpression
+ * for all available types.
+ *
+ * Expressions are evaluated (typically for all rows of a Chunk) using the ExpressionEvaluator.
+ */
+class AbstractExpression : public std::enable_shared_from_this<AbstractExpression> {
+ public:
+  explicit AbstractExpression(const ExpressionType type,
+                              const std::vector<std::shared_ptr<AbstractExpression>>& arguments);
+  virtual ~AbstractExpression() = default;
+
+  /**
+   * Recursively check for Expression equality.
+   * @pre Both expressions need to reference the same LQP
+   */
+  bool deep_equals(const AbstractExpression& other) const;
+
+  virtual bool requires_calculation() const;
+
+  virtual std::shared_ptr<AbstractExpression> deep_copy() const = 0;
+
+  /**
+   * @return a human readable string representing the Expression that can be used as a column name
+   */
+  virtual std::string as_column_name() const = 0;
+
+  /**
+   * @return the DataType of the result of the expression
+   */
+  virtual DataType data_type() const = 0;
+
+  /**
+   * @return whether the result of the Expression MAY contain a NULL
+   */
+  virtual bool is_nullable() const;
+
+  size_t hash() const;
```

Comment:
I find `hash` members more accessible (`exp->hash()` is shorter to type than `std::hash<AbstractExpression>{}(*exp)`) and if we ever need std hash for expressions, then it could just call this method

Reply:
I would prefer if it was that way, too, but I also see a value in conforming to STL standards.

Reply:
Didn't add `struct hash<AbstractExpression>` as we use it nowhere